### PR TITLE
refactor: add edge concept and remove sources

### PR DESCRIPTION
**Pull request by GeorgeMac**
Includes #90

I am going to draft this initially for consideration and discussion.
Also, while I run through and tidy things up and fixup the docs.
However, it is in a state for consideration and discussion.

Screenshot 2024-12-02 at 12 05 22

This brings in a reather large refactor to the concepts.

Primarily, it does the following:

1. Merges the concepts of Phase and Source into just Phase
2. Removes promotion as a responsibility of Phase (Phase are mostly just what Source was)
3. Introduces Edges and in particular the `Promotes` edge implementation

1. Source becomes Phase

Phases before were a generic type over any Source. Now, the Sources have become `core.Phase` implementations themselves. They no longer deal with promotion logic.
Instead, the `core.Phase` abstraction has changed to more of a store (like Source was before) of Resource types.
They expose `Get` and `Update` methods, which are necessary for the new `core.Edge` implementations.

2. Phases no longer promote

As mentioned, now that the sources encapsulate being phases themselves, there is no longer a single `phases.ResourcePhase` with a `Promote` method. This has been completely removed. Instead, this moves onto a new `Edge` implementations.

3. `core.Edge` and `edges.Promote`

Now, instead of phases being something promoted, there are edges between phases that can be _performed_.
An en edge becomes a dependency expression between two phases, which can be performed/invoked.

The `edges` package contains a single edge type called `Promotes`.
This edge type is where our promotion logic has been moved.
Doing this opens the door for future edge kinds (edges now expose their type through this kind property).

Due to this change, the operation is now express on the edge itself, instead of as a button on the nodes in the UI.

Outstanding
- [x] Docs

### CODE REVIEW

**Comment by GeorgeMac** [comment]
Screenshot 2024-12-02 at 13 47 48

I fixed up the funny botton alignment on the edges.

File: pkg/cli/cli.go [hunk 1]
```diff
@@ -123,72 +122,72 @@ func (l labels) Set(v string) error {
...
-		return err
-	}
-
-	if err := promoteAllPhases(ctx, toIter(phase), apply); err != nil {
-		return err
-	}
-
-	return nil
-}
+//func promote(ctx context.Context, s System, args ...string) error {
```
Comment (GeorgeMac) [inline pkg/cli/cli.go:125]: Note to self: restore this (/ how to?)

File: docs/concepts.md [hunk 2]
```diff
@@ -98,56 +105,65 @@ pipeline := glu.NewPipeline(glu.Name("mypipeline"), func() *SomeResource {
...
+	From() Descriptor
+	To() Descriptor
+	Perform(context.Context) (Result, error)
+	CanPerform(context.Context) (bool, error)
 }
 ```
 
+#### Promotion
+
+The core `promotion` kind edge take care of promotion one phase to the next on a call to `Perform(ctx)`.
```
Comment (markphelps) [inline docs/concepts.md:148]:
```suggestion
The core `promotion` kind edge promotes one phase to the next on a call to `Perform(ctx)`.
```

File: docs/concepts.md [hunk 3]
```diff
@@ -98,56 +105,65 @@ pipeline := glu.NewPipeline(glu.Name("mypipeline"), func() *SomeResource {
...
+The process is as follows:
+
+1. Get the current resource state from the source phase based on phase metadata.
+2. Get the current resource state from the destination promotion target phase.
+3. If the resource from (2) is equal to that of (3) (based on comparing their digest), the return (no op).
```
Comment (markphelps) [inline docs/concepts.md:153]:
```suggestion
3. If the resource from (2) is equal to that of (3) (based on comparing their digest), then return (no-op).
```

File: pkg/core/pipeline.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+		edges:  map[string]map[string]Edge{},
+	}
+}
+
+// Metadata returns the metadata assocated with the Pipelines (name and labels).
+func (p *Pipeline) Metadata() Metadata {
+	return p.meta
+}
+
+// Add will add the provided resource phase to the pipeline along with configuring
```
Comment (markphelps) [inline pkg/core/pipeline.go:50]:
```suggestion
// AddPhase will add the provided resource phase to the pipeline along with configuring
```

File: pkg/core/pipeline.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+// IsPhase causes a call to Phases to list specifically the provided phase p.
+func IsPhase(p Phase) containers.Option[PhaseOptions] {
+	return func(co *PhaseOptions) {
+		co.phase = p
+	}
+}
+
+// HasLabel causes a call to Phases to list any phase with the matching name.
```
Comment (markphelps) [inline pkg/core/pipeline.go:202]:
```suggestion
// HasName causes a call to Phases to list any phase with the matching name.
```

File: pkg/edges/edges.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/get-glu/glu/pkg/core"
+)
+
+// Source is an interface around storage for resources.
```
Comment (markphelps) [inline pkg/edges/edges.go:11]:
```suggestion
// Phase is an interface around storage for resources.
```

File: pkg/edges/edges.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+type Phase[R core.Resource] interface {
+	core.Phase
+	GetResource(_ context.Context) (R, error)
+}
+
+// UpdatableSource is a source through which the phase can promote resources to new versions
```
Comment (markphelps) [inline pkg/edges/edges.go:17]:
```suggestion
// UpdatablePhase is a source through which the phase can promote resources to new versions
```

**Review by markphelps — COMMENTED** [review]
couple minor comments to be updated, but otherwise lgtm!

**Pull request merged**